flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceMigrationITCase.java

```diff
@@ -0,0 +1,270 @@
+        final int enumVersion = source.getEnumeratorCheckpointSerializer().getVersion();
+        final int splitVersion = source.getSplitSerializer().getVersion();
+        String testCase = String.format("enum%s-split%s", enumVersion, splitVersion);
+
+        Path savepointPath = KAFKA_SOURCE_SAVEPOINT_PATH.resolve(testCase);
+        Files.createDirectories(savepointPath);
+
+        final CloseableIterator<TimestampedRecord> recordIterator =
+                env.fromSource(source, WatermarkStrategy.noWatermarks(), "TestDataSource")
+                        .uid(KAFKA_SOURCE_UID)
+                        .collectAsync();
+
+        final JobClient jobClient = env.executeAsync();
+        final JobID jobID = jobClient.getJobID();
+
+        final Queue<TimestampedRecord> readRecords = new ConcurrentLinkedQueue<>();
+        while (readRecords.size() < writtenRecords.size() && recordIterator.hasNext()) {
+            readRecords.add(recordIterator.next());
+        }
+
+        try {
+            CompletableFuture<String> savepointFuture =
+                    clusterClient.stopWithSavepoint(
+                            jobID, false, savepointPath.toString(), SavepointFormatType.NATIVE);
+            savepointFuture.get(2, TimeUnit.MINUTES);
+
+            final long maxTS = getMaxTS(writtenRecords);
+            assertThat(readRecords).hasSize(NUM_RECORDS).allMatch(r -> r.getTimestamp() <= maxTS);
+        } finally {
+            SoftAssertions softly = new SoftAssertions();
+            softly.assertThat(clusterClient.requestJobResult(jobID).get())
+                    .returns(ApplicationStatus.SUCCEEDED, JobResult::getApplicationStatus)
+                    .extracting(
+                            JobResult::getSerializedThrowable,
+                            InstanceOfAssertFactories.optional(Exception.class))
+                    .isEmpty();
+            softly.assertAll();
+        }
+    }
+
+    private static long getMaxTS(List<ProducerRecord<String, Integer>> writtenRecords) {
+        return writtenRecords.stream().mapToLong(ProducerRecord::timestamp).max().orElseThrow();
+    }
+
+    private static KafkaSource<TimestampedRecord> createSource() {
+        return KafkaSource.<TimestampedRecord>builder()
+                .setBootstrapServers(KafkaSourceTestEnv.brokerConnectionStrings)
+                .setTopics(TOPIC)
+                .setDeserializer(new TestDeserializer())
+                .build();
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("getKafkaSourceSavepoint")
+    void testRestoreFromSavepointWithCurrentVersion(Path savepointPath) throws Throwable {
+        // this is the part that has been read already in the savepoint
+        KafkaSourceTestEnv.createTestTopic(TOPIC);
+        final List<ProducerRecord<String, Integer>> existingRecords =
+                KafkaSourceTestEnv.getRecordsForTopic(TOPIC);
+        KafkaSourceTestEnv.produceToKafka(existingRecords);
```

Comment:
Let's extract a method to make sure this part is in sync between the setup and the test